contracts/src/libraries/LPMath.sol

```diff
@@ -925,7 +954,60 @@ library LPMath {
             }
         }
 
-        return shareProceeds;
+        // Calculate the present value after applying the share proceeds.
+        bool success_;
+        (
+            params.presentValueParams.shareReserves,
+            params.presentValueParams.shareAdjustment,
+            params.presentValueParams.bondReserves,
+            success_
+        ) = calculateUpdateLiquiditySafe(
+            params.originalShareReserves,
+            params.originalShareAdjustment,
+            params.originalBondReserves,
+            params.presentValueParams.minimumShareReserves,
+            -shareProceeds.toInt256()
+        );
+        if (!success_) {
+            // NOTE: Return 0 to indicate that the share proceeds couldn't be
+            // calculated.
+            return 0;
+        }
+        uint256 presentValue_ = calculatePresentValue(
+            params.presentValueParams
+        );
+
+        // Check to see if the current share proceeds is closer to the optimal
+        // value than the previous closest value. We'll choose whichever of the
+        // share proceeds that is closer to the optimal value.
+        int256 lastDelta = presentValue_.mulDown(lpTotalSupply).toInt256() -
+            params
+                .startingPresentValue
+                .mulUp(params.activeLpTotalSupply)
+                .toInt256();
+        if (lastDelta.abs() < smallestDelta.abs()) {
+            closestShareProceeds = shareProceeds;
+            closestPresentValue = presentValue_;
+        }
+
+        // Verify that the LP share price was conserved within a reasonable
+        // tolerance.
+        if (
+            // NOTE: Round down to make the check stricter.
+            closestPresentValue.divDown(params.activeLpTotalSupply) +
+                SHARE_PROCEEDS_TOLERANCE <
+            params.startingPresentValue.divUp(lpTotalSupply) ||
+            // NOTE: Round up to make the check stricter.
+            closestPresentValue.divUp(params.activeLpTotalSupply) >=
+            params.startingPresentValue.divDown(lpTotalSupply) +
+                SHARE_PROCEEDS_TOLERANCE
```

Comment:
you're using a different tolerance check here (additive / absolute error) than in `shouldShortCircuitDistributeExcessIdleShareProceeds` (multiplicative / relative error):

`PV_1 / l_1 >= (PV_0 / l_0) + tolerance` vs `PV_1 / l_1 >= (PV_0 / l_0) * (1 + tolerance)`.

Reply:
Hmm, yeah that's fair. I'll switch it to multiplicative.